### PR TITLE
Add convenience combinators

### LINKED ISSUE #139: Include batteries
One of Hedgehog's main strengths, IMHO, is its flexibility due to all the `Gen` combinators. However, practical use entirely depends on the user to define commonly used combinations themselves to avoid too much tedious repetition in the tests, and then either repeat these definitions for each test project, or create a separate solution/assembly with common usecases and add that to all test projects. I think most people will just copy-paste the same common combinations across all of their test projects (copy-paste).

I'd like Hedgehog to have a bit more "batteries included" approach. This would make the framework a much easier sell, so to speak. This could either be done by including common shortcuts in Hedgehog itself, or it could be released as a separate nuget (Hedgehog.Batteries?) maintained by Hedgehog maintainers or someone else. If it's a separate library maintained by someone else, it should preferably be officially endorsed by Hedgehog.

Here are examples of what I'm talking about (I'm not saying everything here must be included):

```f#
namespace Hedgehog.Batteries

module Gen =

  open Hedgehog

  /// Generates an int using sensible defaults (Range.exponentialBounded)
  let dInt = Gen.int <| Range.exponentialBounded()

  /// Generates a string using sensible defaults (Range.exponential 0 50, latin1)
  let dString = Gen.string (Range.exponential 0 50) Gen.latin1

  /// Generates a list using sensible defaults (Range.exponential 0 50)
  let dList g =
    g |> Gen.list (Range.exponential 0 50)

  /// Generates a list with length determined using Range.exponential lower upper
  let eList lower upper =
    Gen.list (Range.exponential lower upper)

  /// Generates a list with length determined using Range.linear lower upper
  let lList lower upper =
    Gen.list (Range.linear lower upper)

  /// Generates a list with length determined using Range.constant lower upper
  let cList lower upper =
    Gen.list (Range.constant lower upper)

  /// Generates null part of the time
  let withNull g =
    g |> Gen.option |> Gen.map (fun xOpt ->
      match xOpt with Some x -> x | None -> null)

  /// Generates a value that is not null
  let noNull g =
    g |> Gen.filter (not << isNull)

  /// Generates a value that is not equal to another value
  let notEqual other =
    Gen.filter ((<>) other)

  /// Generates a value that is not equal to another option-wrapped value
  let notEqualOpt other =
    Gen.filter (fun x -> match other with Some o -> x <> o | None -> true)


  /// Generates a 2-tuple with sorted elements.
  let sorted2 g =
    g |> Gen.map (fun (x1, x2) -> 
           let l = [x1; x2] |> List.sort
           (l.Item 0, l.Item 1))

  /// Generates a 3-tuple with sorted elements.
  let sorted3 g =
    g |> Gen.map (fun (x1, x2, x3) -> 
           let l = [x1; x2; x3] |> List.sort
           (l.Item 0, l.Item 1, l.Item 2))

  /// Generates a 4-tuple with sorted elements.
  let sorted4 g =
    g |> Gen.map (fun (x1, x2, x3, x4) -> 
           let l = [x1; x2; x3; x4] |> List.sort
           (l.Item 0, l.Item 1, l.Item 2, l.Item 3))

  /// Generates a 2-tuple with distinct elements.
  let distinct2 g =
    g |> Gen.filter (fun (x1, x2) -> x1 <> x2)

  /// Generates a 3-tuple with distinct elements.
  let distinct3 g =
    g |> Gen.filter (fun (x1, x2, x3) -> 
           [x1; x2; x3] |> List.distinct = [x1; x2; x3])

  /// Generates a 4-tuple with distinct elements.
  let distinct4 g =
    g |> Gen.filter (fun (x1, x2, x3, x4) -> 
           [x1; x2; x3; x4] |> List.distinct = [x1; x2; x3; x4])

  /// Generates a 2-tuple with strictly increasing elements
  let increasing2 g =
    g |> sorted2 |> distinct2

  /// Generates a 3-tuple with strictly increasing elements
  let increasing3 g =
    g |> sorted3 |> distinct3

  /// Generates a 4-tuple with strictly increasing elements
  let increasing4 g =
    g |> sorted4 |> distinct4

  /// Generates a tuple of datetimes where the range determines the number
  /// of days apart (positive numbers -> increasing order and vice versa).
  let dateInterval dayRange =
    gen {
      let tickRange = 
        dayRange 
        |> Range.map (fun days -> int64 days * System.TimeSpan.TicksPerDay)
      let! ticksApart = Gen.integral tickRange
      let! dt1 = Gen.dateTime |> Gen.filter (fun dt -> 
        dt.Ticks + ticksApart > System.DateTime.MinValue.Ticks
        && dt.Ticks + ticksApart < System.DateTime.MaxValue.Ticks)
      let dt2 = dt1.AddTicks ticksApart
      return dt1, dt2
    }


  /// Generates a list using inpGen together with a function that maps each of
  /// the distinct elements in the list to values generated by outGen. Distinct
  /// elements in the input list may possibly map to the same values. The generated
  /// function throws if called with values not in the input list.
  let withMapTo (outGen:Gen<'b>) (inpGen:Gen<'a list>) : Gen<'a list * ('a->'b)> =
    gen {
      let! inputs = inpGen
      let! outputs = outGen |> Gen.list (Range.singleton inputs.Length)
      let inOutMap = List.zip inputs outputs |> Map.ofList
      return inputs, (fun x -> inOutMap.Item x)
    }

  /// Generates a list using inpGen together with a function that maps each of
  /// the distinct elements in the list to values generated by outGen. Distinct
  /// elements in the input list are guaranteed to map to distinct values. Only
  /// use this if the output space is large enough that a list of distinct output
  /// values is likely to be generated. The generated function throws if called
  /// with values not in the input list.
  let withDistinctMapTo (outGen:Gen<'b>) (inpGen:Gen<'a list>) : Gen<'a list * ('a->'b)> =
    gen {
      let! inputs = inpGen
      let! outputs =
        outGen
        |> Gen.list (Range.singleton inputs.Length)
        |> Gen.filter (fun lst -> lst = List.distinct lst)
      let inOutMap = List.zip inputs outputs |> Map.ofList
      return inputs, (fun x -> inOutMap.Item x)
    }

  /// Inserts the given element at a random place in the list
  let addElement (x:'a) (g:Gen<'a list>) =
    gen {
      let! xs = g
      let! i = Gen.integral (Range.constant 0 xs.Length)
      let l1, l2 = xs |> List.splitAt i
      return List.concat [l1; [x]; l2]
    }
```

### PR DESCRIPTION
**Pull request by cmeeren**
Closes #139.

In addition to combinators mentioned in #139, I decided to add several others I have found helpful. Let me know if any should be removed.

### CODE REVIEW

**Comment by moodmosaic** [comment]
The additional ones you’ve added are interesting, but it’d allow us to move faster if you could propose them on a separate PR.

It will also be easier for @jystic to follow up on this PR if we keep it focused on those combinations we discussed in #139.

Thank you for all the your hard work, @cmeeren!

**Comment by cmeeren** [comment]
> The additional ones you’ve added are interesting, but it’d allow us to move faster if you could propose them on a separate PR.

Sure, I can do that, but personally I'm in no hurry. Do you think it's necessary, or could we just make the decision on each combinator in this PR?

Also, the new improved implementation of `withDistinctMapTo` makes use of the new `notIn` combinator, so either that'd need to be included, or it would need to be manually coded into the `withDistinctMapTo` implementation. And if `notIn` is to be included, then `notContains` is a natural companion. And then we're very close to `notSubstringOf` which is kind of the same only for strings, and so on and so forth.

Still, if you want me to remove some/all of the ombinators not mentioned in #139 and include them in a separate PR after this is merged, I'll make it happen. Let me know how you want to proceed.

File: src/Hedgehog/Gen.fs [hunk 1]
```diff
@@ -447,6 +447,267 @@ module Gen =
...
+      filter <| fun s ->
+          str.IndexOf(s, System.StringComparison.OrdinalIgnoreCase) = -1
+
+    /// Generates a string that does not start with another string.
+    let notStartsWith (str : string) : (Gen<string> -> Gen<string>) =
+      filter <| fun s -> not <| s.StartsWith str
+
+    /// Generates a string that does not start with another string using
+    /// StringComparison.OrdinalIgnoreCase.
+    let iNotStartsWith (str : string) : (Gen<string> -> Gen<string>) =
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Perhaps `iNotStartsWith` a bit cryptic 😕 Could we come up with a more descriptive name? 😕

File: src/Hedgehog/Gen.fs [hunk 2]
```diff
@@ -447,6 +447,267 @@ module Gen =
...
+        filter <| fun s ->
+            not <| str.Equals(s, System.StringComparison.OrdinalIgnoreCase)
+
+    /// Generates a string that is not a substring of another string.
+    let notSubstringOf (str : string) : (Gen<string> -> Gen<string>) =
+      filter <| fun s -> not <| str.Contains s
+
+    /// Generates a string that is not a substring of another string using
+    /// StringComparison.OrdinalIgnoreCase.
+    let iNotSubstringOf (str : string) : (Gen<string> -> Gen<string>) =
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Perhaps `iNotSubstringOf` is a bit cryptic 😕 Could we come up with a more descriptive name? 😕

File: src/Hedgehog/Gen.fs [hunk 3]
```diff
@@ -447,6 +447,267 @@ module Gen =
...
+        g |> filter (fun x -> not <| List.contains x list)
+
+    /// Generates a list that does not contain the specified element.
+    /// Shortcut for Gen.filter (not << List.contains x)
+    let notContains (x: 'a) : (Gen<'a list> -> Gen<'a list>) =
+      filter (not << List.contains x)
+
+    /// Generates a string that is not equal to another string using
+    /// StringComparison.OrdinalIgnoreCase.
+    let iNotEqualTo (str : string) : (Gen<string> -> Gen<string>) =
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: Perhaps `iNotEqualTo` is a bit cryptic 😕 Could we come up with a more descriptive name? 😕

File: src/Hedgehog/Gen.fs [hunk 4]
```diff
@@ -447,6 +447,267 @@ module Gen =
...
+        let randomizeCharCase c =
+            let f = if r.Next() % 2 = 0
+                    then System.Char.ToUpper
+                    else System.Char.ToLower
+            f c
+
+        s |> Seq.map randomizeCharCase |> System.String.Concat
+
+    /// Randomizes the case of the characters in the string.
+    let casePermutationOf (str : string) : Gen<string> =
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: This one uses `randomizeCase` which in turn uses `System.Random ()` under the cover, which we should consider using (or not using) once we're done with #135.

For this reason, I'd like to propose to discuss about it separately, on next PR(s), and not include it on this one.

File: src/Hedgehog/Gen.fs [hunk 5]
```diff
@@ -447,6 +447,267 @@ module Gen =
...
+            arr.[i] <- arr.[j]
+            arr.[j] <- item
+        let rnd = new System.Random()
+        let ln = arr.Length
+        [0..(ln - 2)] |> Seq.iter (fun i -> swap i (rnd.Next(i, ln)))
+        arr |> Array.toList
+
+    /// Generates a permutation the specified list (shuffles its elements).
+    let permutationOf (list : 'a list) : Gen<'a list> =
+        gen { return list |> shuffle }
```
Comment (moodmosaic) [inline src/Hedgehog/Gen.fs]: This one uses `shuffe` which in turn uses `System.Random ()` under the cover, which we should consider using (or not using) once we're done with #135.

For this reason, I'd like to propose to discuss about it separately, on next PR(s), and not include it on this one.

**Comment by cmeeren** [comment]
> I suggest we take on separately: ... 

Sure, makes sense. I'll remove them from this PR.

> And perhaps we can find better names for i...

We could call it `notSubstringOfIgnoringCase` etc., but that's awfully long. I can't really think of anything shorter. I'm used case-insensitive functions being prefixed with `i`, so for me `iNotSubstringOf` isn't a problem. Also, I'd guess users wanting this functionality are probably writing `Gen.substring` to see which functions turn up in Intellisense, and the function documentation makes it immediately clear what it does.

Let me know if you have a better name; I can't think of any that does not use `i` and is not awfully long-winded.

**Comment by moodmosaic** [comment]
>Let me know if you have a better name; I can't think of any that does not use `i` and is not awfully long-winded.

I can't currently come up with a better name...

In fact, the other option, `notSubstringOfIgnoringCase`, hints that we're doing 2 things in 1 call, and perhaps we should refactor 1 of those 2 things as a parameter...

Still, we can discuss as much as we can on separate PR(s), so instead of jumping the gun by 'forcing' a name we're not entirely sure about, we can move all those `iNotXyz` ones on a separate PR...

The more short and focused a PR is, the easier it comes for reviewers to find a continuous time-slot to review it, try it, and perhaps even merge it. (Some interesting tips [here](http://blog.ploeh.dk/2015/01/15/10-tips-for-better-pull-requests/).)

**Comment by cmeeren** [comment]
Thank you for the link, Mark makes good sense as always. I have now removed both the `i...` and its corresponding case-sensitive brethren.

**Comment by cmeeren** [comment]
One thing I'd like to do is add unit tests for some of the combinators. However, I can't see how this can be done sensibly without using property-based testing and auto-generated values. How do you feel about using basic Hedgehog generators/combinators to test the more advanced ones?

**Comment by moodmosaic** [comment]
If you want to add automated tests, there are a few ways to do this:

* Add example(s) in the Src/Script.fsx
* Add example-based (traditional) tests using xUnit.net in the Hedgehog.Tests project
* Add property-based tests using Hedgehog in the Hedgehog.Tests project which already references Hedgehog anyway.

**Comment by moodmosaic** [comment]
Up to now, we usually add examples in the Src/Script.fsx...

**Comment by cmeeren** [comment]
> Up to now, we usually add examples in the Src/Script.fsx...

As far as I can see, that serves only to exemplify, not test for correct behavior.

I've added some tests now to `GenTests.fs`.

Also, I'm wondering if there really is much utility for `cArray`, `cSeq` , etc. (since one can always use `cList` etc. together with `Array.fromList` etc., and in my experience, I mostly only use lists in tests). What do you think, should I remove the six `xArray`/`xSeq` functions?

**Comment by moodmosaic** [comment]
The more you remove the better; because once you add them it's *breaking* change to remove something.

Yes, we're on version < 1, and according to Semantic Versioning 2.0, breaking changes are allowed... Still, we should try to conservative when possible.

**Comment by cmeeren** [comment]
True that. All right, I've removed some more now. It's very close to #139; the only additions are `eString`, `lString`, `cString`. `notIn`, and `notContains`. How do you feel about that? :)

**Comment by cmeeren** [comment]
PS, don't feel you need to cater to my whims. If you want those too removed from this PR, I'll do that.

**Comment by moodmosaic** [comment]
Those look fine to me 👍

**Comment by moodmosaic** [comment]
@cmeeren, should we close this one in in favor of https://github.com/hedgehogqa/fsharp-hedgehog-experimental/pull/4?

**Comment by cmeeren** [comment]
Certainly.

**Pull request closed without merging**